decode_test.go

```diff
@@ -1241,12 +1249,58 @@ c:
 			actual := fmt.Sprintf("%+v", value.Elem().Interface())
 			expect := fmt.Sprintf("%+v", test.value)
 			if actual != expect {
+				if deepCompare(reflect.ValueOf(test.value), value.Elem()) {
+					return
+				}
 				t.Fatalf("failed to test [%s], actual=[%s], expect=[%s]", test.source, actual, expect)
 			}
 		})
 	}
 }
 
+// deepCompare
+func deepCompare(v1, v2 reflect.Value) bool {
+	if v1.Kind() != v2.Kind() {
+		return false
+	}
+
+	switch v1.Kind() {
+	case reflect.Map:
+		if v1.Len() != v2.Len() {
+			return false
+		}
+
+		for _, k1 := range v1.MapKeys() {
+			found := false
+			for _, k2 := range v2.MapKeys() {
+				if getPtr(k1) == getPtr(k2) {
+					if getPtr(v1.MapIndex(k1)) != getPtr(v2.MapIndex(k2)) {
+						return false
+					}
+					found = true
+					break
+				}
+			}
+			if !found {
+				return false
+			}
+		}
+		return true
+	}
+	return false
+}
```

Comment:
There is a lot of branches, here is my suggestion

```suggestion
func deepCompare(v1, v2 reflect.Value) bool {
	if v1.Kind() != v2.Kind() || v1.Kind() != reflect.Map || v1.Len() != v2.Len() {
		return false
	}

	for _, k1 := range v1.MapKeys() {
		for _, k2 := range v2.MapKeys() {
			if getPtr(k1) != getPtr(k2) {
				return false
			}

			if getPtr(v1.MapIndex(k1)) != getPtr(v2.MapIndex(k2)) {
				return false
			}
	}
	return true
}
```